src/nvim/ex_getln.c

```diff
@@ -2263,20 +2280,56 @@ static void draw_cmdline(int start, int len)
     msg_outtrans_len(ccline.cmdbuff + start, len);
 }
 
+void ui_ext_cmdline_char(int c, int shift)
```

Comment:
the pattern "certain classes of code must be in a separate function, even if it is very short and integral with the surrounding code" I have never heard of.

OTOH, what could make sense is generating marshaling wrappers, to avoid manually contructing the array. But there will still be external ui specific code to prepare the values.